Added SubArray tests for linalg/matmul.jl and fixed Vector -> StridedVector in dot

pkofod opened the pull request:
<s>First of all: I know there are many commits. They are to be squashed, but they are there for my own sake until this is (hopefully) ready to be merged.</s>

I've generally tried to do what I did in #15364, #15354, #15314, but I tried not to close my eyes and add `for Atype in ["Array", "SubArray"]` everywhere without thinking. Some cases I ignored were the ones already testing `SubArray` related problems, and also the `gemv!` tests. The latter was mostly due to the fact that `gemv!` is defined for `Strided*`, and the number of combinations of matrices of `Array` and `SubArray` is large. My concern was that these tests would take much longer to run, but for no real reason. I would love comments on this.

I also added `let` blocks around groups of tests with named matrices and vectors (`A`, `B`, ..., `b`, ...). This is because running all the tests in `julia/test/` defines a lot of variables that can potentially be used by mistake if all contributors are not equally careful or lucky. At some point I had removed a `A =` definition somewhere, and another `A` was used by mistake (and didn't cause a test fail by chance). This is not ideal, as there's a chance such a mistake is not caught. If people don't like them, I can remove them again.

Other points to be made
- The very first commit removes a 10 times loop, bit `i` is never used, so I removed it. Was there a reason simply couldn't infer from the test?
- In commit 4, `vf` was constructed, but an old `v` was used
- <s>fixed `dot` such that it is defined for `StridedVector` instead of `Vector`</s> This was not a good idea, and has been reverted.

pkofod committed:
Add SubArray tests in matmul.

tkelman commented on test/linalg/matmul.jl:
```diff
@@ -5,189 +5,279 @@ using Base.Test
 ## Test Julia fallbacks to BLAS routines
 
 # matrices with zero dimensions
-for i = 1:10
```
blame says @andreasnoack https://github.com/JuliaLang/julia/commit/687d2b0da6c5e8cff895e8c6f7f8e4b6866b7997 - maybe because `Array` would be uninitialized in the non-zero-dimensional case?

pkofod replied:
Okay. I'm not quite sure if I understand why you would have to loop ten times in that case either, but I'll look at it again.

andreasnoack replied:
I don't remember why I did that (a comment would probably have helped), but my guess is what @tkelman said. When multiplying an mx0 matrix with an 0xn then we have a fast branch that just allocates an mxn matrix and sets the elements to zero. Repeating the tests makes it more likely to catch it if we forgot to set the memory to zero.

pkofod replied:
Okay, I get it. Do you want a loop again then, but with a comment?

andreasnoack replied:
I think it fine to remove the loop. We should see occational test failures even without it if the arrays are not set correctly.

andreasnoack commented on base/linalg/matmul.jl:
```diff
@@ -44,7 +44,7 @@ end
 vecdot{T<:BlasComplex}(x::Union{DenseArray{T},StridedVector{T}}, y::Union{DenseArray{T},StridedVector{T}}) = BLAS.dotc(x, y)
 
-function dot{T<:BlasReal, TI<:Integer}(x::Vector{T}, rx::Union{UnitRange{TI},Range{TI}}, y::Vector{T}, ry::Union{UnitRange{TI},Range{TI}})
+function dot{T<:BlasReal, TI<:Integer}(x::StridedVector{T}, rx::Union{UnitRange{TI},Range{TI}}, y::StridedVector{T}, ry::Union{UnitRange{TI},Range{TI}})
```
The `dot` definitions are a bit too scattered and I think there is a similar definition in `blas.jl`. Could you check how they much overlap?

pkofod replied:
The closest I can find is https://github.com/JuliaLang/julia/blob/master/base/linalg/blas.jl#L202 which is without the `Range` positional arguments.

andreasnoack replied:
Maybe we shouldn't allow both that the vectors are strided and a range argument. It seems kind of double to me. Ideally these methods should actually be deleted and we should just use subarrays but there is probably still a price to pay over using this approach.

pkofod replied:
Additionally, isn't the following 
```julia
function dot{T<:BlasReal}(DX::Union{DenseArray{T},StridedVector{T}}, DY::Union{DenseArray{T},StridedVector{T}})
```
rather strange, as `DenseArray <: StridedVector == true`? I mean, wouldn't `StridedVector` suffice?

andreasnoack replied:
No `DenseArray` is not a subtype of `StridedVector. The reason we have this is to be able to handle e.g. a matrix as a vector without explicitly reshaping it. I think it would be cleaner if we could simply reshape without cost. I'm now sure how often people use this feature for `dot` but it was requested for `axpy` at some point.

pkofod replied:
Ah, sorry, I missed the Array, even if I typed it myself.
> Maybe we shouldn't allow both that the vectors are strided and a range argument. It seems kind of double to me.
You mean in the change I made? I can see you point, but it's been a while. I'll have to look at it again, and see if I can figure out how to address your point.

pkofod replied:
@andreasnoack Okay, so I thought about it. Maybe I should just revert the definition, and only run the `dot(vector, range, vector, range)` method for `Array`s, and not `SubArray`s. This will mean that `dot(a::SubArray, b::Range, c::SubArray, d::Range)` won't have a method defined, but maybe that actually makes sense as you hinted at above. Why not just create a different sub with the appropriate shape if that is what you want. I'll make this change, and you can comment if you agree or not.

pkofod replied:
> Maybe we shouldn't allow both that the vectors are strided and a range argument. It seems kind of double to me. Ideally these methods should actually be deleted and we should just use subarrays but there is probably still a price to pay over using this approach.

@andreasnoack Do you mean sub-ing according to the input ranges, and calling the `StridedVector` version of `dot(c)`? I tried that, but there's a substantial relative time increase. https://gist.github.com/pkofod/901639371f4c05925e171dd279147f16

andreasnoack commented:
@pkofod Please rebase when you have the time. Except for the `dot` comment, I think this ready.

pkofod commented:
@andreasnoack tests pass; should I squash to one commit?

andreasnoack commented:
I think two commits are fine here, but please see my reply to your reply to my comment regarding the `dot` method.

pkofod commented:
Rebased and pending tests. I reverted the changes to `dot`, as it didn't make sense. The point was to make the tests under the line `# issue #6450` run for `SubArray`s, but it is not obvious that we want those functions to be tested for `SubArray`s. Instead, `dot` should be called without the ranges, and a new sub can be created if different ranges are wanted (instead of using `dot(vec, range, vec, range)`).

andreasnoack commented:
Yes. I think that is the right way to do it. Thanks.
